### DISCUSSION TITLE
Topic: Unexpected 301 Permanent Redirects for Prefix-Based Routes in Next.js 14.2.5

### DISCUSSION THREAD

**Discussion by KamaniBhavin** [vercel/next.js#73126] [App Router] [answered]
**Description:**

In a Next.js 14.2.5 application, I encountered an issue where defining two similarly prefixed routes, `/register-kit` and `/register-store`, resulted in unexpected **301 permanent redirects** to `/register`. The `/register` route did not exist, but Next.js still redirected requests to it.

Here’s a detailed explanation of the problem, my observations, and how I eventually resolved it by restructuring the routes.

---

### **Steps to Reproduce**

1. **Setup File-Based Routing:**
   In the `app` directory, define the following two routes:
   ```
   app/register-kit/page.tsx
   app/register-store/page.tsx
   ```

2. **Navigate to Routes:**
   Open the browser and try accessing `/register-kit` or `/register-store`.

3. **Observe the Redirection:**
   Both `/register-kit` and `/register-store` redirected to `/register`, even though `/register` does not exist in the file system.

---

### **Observed Behavior**

- A **301 Permanent Redirect** occurred for both `/register-kit` and `/register-store`, leading to `/register`.
- The browser's developer tools showed the following response headers:
  ```
  x-edge-runtime: 1
  x-matched-path: /register
  x-powered-by: Next.js
  ```
  This indicated that Next.js was incorrectly resolving the path to `/register`.

---

### **Workaround**

To resolve the issue, I changed the route structure to use nested paths, as follows:
```
app/register/kit/page.tsx
app/register/store/page.tsx
```

After making this change:
- `/register/kit` and `/register/store` worked correctly.
- There were no more 301 permanent redirects, and the correct pages were served.

---

### **Expected Behavior**

1. Routes like `/register-kit` and `/register-store` should work independently without requiring a nested structure.
2. If there is a conflict or ambiguity in the routing logic, Next.js should:
   - Provide clear warnings or logs during `next dev` or `next build`.
   - Avoid returning a 301 redirect to a non-existent route (`/register` in this case).

---

### **Feedback and Questions**

1. **Is this Expected Behavior?**
   - Should similarly prefixed routes like `/register-kit` and `/register-store` require a nested structure to avoid conflicts?

2. **Improved Debugging Tools**
   - Could Next.js provide clearer warnings or error messages during `next dev` or `next build` for such routing conflicts?
   - For example, when Next.js resolves `/register-kit` to `/register`, a warning could indicate a potential conflict or ambiguity.

3. **Routing Resolution Enhancements**
   - Are there plans to make routing resolution more robust, especially for prefix-based routes that could lead to implicit conflicts?

---

### **Environment**

- **Next.js Version:** 14.2.5
- **Node.js Version:** `18.20`

---

**Key Observations:**

- The issue is entirely resolved when switching to a nested structure (`/register/kit` and `/register/store`).
- However, the 301 permanent redirect behavior when `/register-kit` and `/register-store` are used independently seems unexpected and counterintuitive.

Would love to hear from others if they’ve faced similar issues and how they addressed them. Additionally, any insights or guidance from the Next.js maintainers on whether this is a known limitation or a bug would be greatly appreciated!

**Comment by denis23x** [comment, marked as answer]
Hi @KamaniBhavin 

I tried to reproduce your problem, and there is no problem in my case 🤷‍♂️

Probably you are using custom middleware or external packages which affecting your routing.

<img width="418" alt="image" src="https://github.com/user-attachments/assets/f136236b-8562-42d2-ad8b-c29bda0db482">

<img width="418" alt="image" src="https://github.com/user-attachments/assets/3185108f-c55b-4ee9-a786-48684a8bb19e">

<img width="783" alt="image" src="https://github.com/user-attachments/assets/af86df35-165d-4a15-b10f-9ac24868059b">

<img width="214" alt="image" src="https://github.com/user-attachments/assets/5bc7e4ed-e788-4bf5-a63a-d1c003070293">

**Reply by KamaniBhavin** [reply]
@denis23x Thanks for taking the time to look into this! I'm still a bit puzzled about what caused the issue. Initially, I didn't think it was a bug, which is why I started a discussion about it. As far as I know, I do have middleware in place, but even when I bypassed it, the issue persisted. It’s possible that some other package could be interfering. 

The strangest part is that everything worked fine in my local and DEV environments, but as soon as it was deployed to PROD, things got buggy. 

Anyway, I really appreciate you giving it a shot and offering your input. I'm pretty confident this isn't a core problem. Cheers to the Next.js developers community for being so active and supportive! 🚀
